File: squeeze_score.py

from typing import Dict, Iterable, List, Union
import sys, csv, argparse, io

FLOAT_BUCKETS = [
    (5_000_000, 35),
    (10_000_000, 30),
    (20_000_000, 25),
    (50_000_000, 15),
    (float("inf"), 5),
]

REGSHO_BONUS = 4

def apply_minimums_additive(score: float, float_shares: float, ctb: float, si_pct: float) -> float:
    if float_shares < 10_000_000 and ctb > 300:
        score = max(score, 91)
    if float_shares < 5_000_000 and ctb > 100:
        score = max(score, 88)
    if float_shares < 30_000_000 and ctb > 50 and si_pct >= 20:
        score = max(score, 85)
    return score

def apply_minimums_multiplicative(score: float, float_shares: float, ctb: float, si_pct: float) -> float:
    if float_shares < 10_000_000 and ctb > 300:
        score = max(score, 91)
    if float_shares < 5_000_000 and ctb > 100:
        score = max(score, 88)
    if float_shares < 30_000_000 and ctb > 50 and si_pct >= 20:
        score = max(score, 85)
    return score

def float_multiplier(float_shares: float) -> float:
    if float_shares < 5_000_000: return 3.0
    if float_shares < 10_000_000: return 2.5
    if float_shares < 20_000_000: return 2.0
    if float_shares < 50_000_000: return 1.5
    return 1.0

def ctb_multiplier(ctb: float) -> float:
    if ctb >= 500: return 3.0
    if ctb >= 300: return 2.7
    if ctb >= 100: return 2.0
    if ctb >= 50: return 1.6
    if ctb >= 25: return 1.3
    return 1.0

def si_multiplier(si_pct: float) -> float:
    if si_pct >= 50: return 2.5
    if si_pct >= 20: return 2.0
    if si_pct >= 10: return 1.5
    if si_pct >= 5: return 1.2
    return 1.0

def ftd_multiplier(ftd_val: float) -> float:
    if ftd_val >= 10_000_000: return 1.3
    if ftd_val >= 5_000_000: return 1.2
    if ftd_val >= 1_000_000: return 1.1
    return 1.0

def score_additive(float_shares: float, si_pct: float, ctb: float, ftd_val: float, regsho: bool) -> float:
    score = 0.0
    if float_shares < 5_000_000: score += 35
    elif float_shares < 10_000_000: score += 30
    elif float_shares < 20_000_000: score += 25
    elif float_shares < 50_000_000: score += 15
    else: score += 5

    if ctb >= 500: score += 24
    elif ctb >= 300: score += 23
    elif ctb >= 100: score += 19
    elif ctb >= 50: score += 14
    elif ctb >= 25: score += 10
    elif ctb >= 10: score += 5

    if si_pct >= 50: score += 25
    elif si_pct >= 20: score += 19
    elif si_pct >= 10: score += 14
    elif si_pct >= 5: score += 10

    if float_shares < 10_000_000 and si_pct >= 20: score += 8
    if float_shares < 5_000_000 and ctb > 100: score += 12
    if float_shares < 10_000_000 and ctb > 300: score += 8

    if ftd_val >= 10_000_000: score += 9
    elif ftd_val >= 5_000_000: score += 7
    elif ftd_val >= 1_000_000: score += 5

    if regsho: score += REGSHO_BONUS

    score = apply_minimums_additive(score, float_shares, ctb, si_pct)
    return min(score, 99.0)

def score_multiplicative(float_shares: float, si_pct: float, ctb: float, ftd_val: float, regsho: bool) -> float:
    raw = (
        float_multiplier(float_shares) *
        ctb_multiplier(ctb) *
        si_multiplier(si_pct) *
        ftd_multiplier(ftd_val) *
        (1.05 if regsho else 1.0)
    )
    score = (raw / 45.0) * 100.0
    score = apply_minimums_multiplicative(score, float_shares, ctb, si_pct)
    return min(score, 99.0)

def score_consensus(float_shares: float, os_shares: float, si_pct: float, ctb: float, ftd_val: float, regsho: bool) -> float:
    add_ = score_additive(float_shares, si_pct, ctb, ftd_val, regsho)
    mul_ = score_multiplicative(float_shares, si_pct, ctb, ftd_val, regsho)
    return round((add_ + mul_) / 2.0, 1)

def score_breakdown(float_shares: float, os_shares: float, si_pct: float, ctb: float, ftd_val: float, regsho: bool) -> Dict[str, float]:
    float_pts = 35 if float_shares < 5_000_000 else 30 if float_shares < 10_000_000 else 25 if float_shares < 20_000_000 else 15 if float_shares < 50_000_000 else 5
    ctb_pts = 24 if ctb >= 500 else 23 if ctb >= 300 else 19 if ctb >= 100 else 14 if ctb >= 50 else 10 if ctb >= 25 else 5 if ctb >= 10 else 0
    si_pts = 25 if si_pct >= 50 else 19 if si_pct >= 20 else 14 if si_pct >= 10 else 10 if si_pct >= 5 else 0

    bonus = 0
    if float_shares < 10_000_000 and si_pct >= 20: bonus += 8
    if float_shares < 5_000_000 and ctb > 100: bonus += 12
    if float_shares < 10_000_000 and ctb > 300: bonus += 8

    ftd_pts = 9 if ftd_val >= 10_000_000 else 7 if ftd_val >= 5_000_000 else 5 if ftd_val >= 1_000_000 else 0
    reg_pts = 4 if regsho else 0

    add_total = float_pts + ctb_pts + si_pts + bonus + ftd_pts + reg_pts
    add_total = apply_minimums_additive(add_total, float_shares, ctb, si_pct)
    add_total = min(add_total, 99.0)

    mul_total = score_multiplicative(float_shares, si_pct, ctb, ftd_val, regsho)
    consensus = round((add_total + mul_total) / 2.0, 1)

    return {
        "Float": float_pts,
        "CTB": ctb_pts,
        "SI%": si_pts,
        "Bonuses": bonus,
        "FTD": ftd_pts,
        "RegSHO": reg_pts,
        "Additive Total": round(add_total, 1),
        "Multiplicative Total": round(mul_total, 1),
        "Consensus": consensus,
    }

# ---------- CSV CLI ----------
def _to_bool(v: Union[str, int, float, bool]) -> bool:
    if isinstance(v, bool): return v
    s = str(v).strip().lower()
    return s in {"1","true","t","yes","y"}

def _float_or_zero(v) -> float:
    try:
        return float(str(v).replace(",","").strip())
    except:
        return 0.0

HEADER_ALIASES = {
    "ticker": {"ticker","symbol"},
    "float_shares": {"float_shares","free float","float","free_float"},
    "os_shares": {"os_shares","shares outstanding","outstanding","os","shares_outstanding"},
    "si_pct": {"si_pct","short float %","short %","short_percent","short_float_pct"},
    "ctb": {"ctb","cost to borrow","borrow cost","ctb%","cost_to_borrow"},
    "ftd_val": {"ftd_val","ftd $","ftd value","ftd","ftd_usd"},
    "regsho": {"regsho","reg sho","threshold","on_regsho"},
}

def _normalize_headers(fieldnames):
    # Build a mapping from canonical -> actual header name in the file
    fieldnames = [fn.strip() for fn in (fieldnames or [])]
    lower_map = {fn.lower(): fn for fn in fieldnames}
    mapping = {}
    for canon, aliases in HEADER_ALIASES.items():
        for a in aliases:
            if a.lower() in lower_map:
                mapping[canon] = lower_map[a.lower()]
                break
    return mapping

def main():
    ap = argparse.ArgumentParser(description="Short-squeeze scoring CLI (robust)")
    ap.add_argument("input", nargs="?", default="-", help="input CSV (default: stdin)")
    ap.add_argument("-o", "--output", default="-", help="output CSV (default: stdout)")
    ap.add_argument("--debug", action="store_true", help="print row-level errors to stderr")
    args = ap.parse_args()

    fin = sys.stdin if args.input == "-" else open(args.input, "r", newline="", encoding="utf-8-sig")
    data = fin.read()
    if fin is not sys.stdin: fin.close()

    # Sniff dialect to support semicolons/tabs
    try:
        dialect = csv.Sniffer().sniff(data.splitlines()[0] + "\n" + data.splitlines()[1])
    except Exception:
        dialect = csv.getdialect("excel")
    reader = csv.DictReader(io.StringIO(data), dialect=dialect)

    header_map = _normalize_headers(reader.fieldnames)
    out_fields = ["ticker","Consensus","Additive Total","Multiplicative Total","Float","CTB","SI%","Bonuses","FTD","RegSHO"]
    fout = sys.stdout if args.output == "-" else open(args.output, "w", newline="", encoding="utf-8")
    writer = csv.DictWriter(fout, fieldnames=out_fields)
    writer.writeheader()

    idx = 0
    for row in reader:
        idx += 1
        try:
            def get(canon, default=""):
                actual = header_map.get(canon)
                return row.get(actual, default) if actual else default

            ticker = get("ticker","")
            float_shares = _float_or_zero(get("float_shares", 0))
            os_shares = _float_or_zero(get("os_shares", 0))
            si_pct = _float_or_zero(get("si_pct", 0))
            ctb = _float_or_zero(get("ctb", 0))
            ftd_val = _float_or_zero(get("ftd_val", 0))
            regsho = _to_bool(get("regsho", False))

            bd = score_breakdown(float_shares, os_shares, si_pct, ctb, ftd_val, regsho)
            bd["ticker"] = ticker
            writer.writerow(bd)
        except Exception as e:
            if args.debug:
                print(f"[row {idx}] error: {e} | row={row}", file=sys.stderr)

    if fout is not sys.stdout: fout.close()

if __name__ == "__main__":
    main()

File: squeezeScore.js
// squeezeScore.js — ES module for Chrome extensions / web
// Nothing hits 100; scores capped at 99.

/** ---------- Utilities ---------- */
const clamp99 = (x) => Math.min(x, 99);
const asFloat = (v) => {
  if (v == null) return 0;
  const s = String(v).trim().replace(/,/g, "");
  const n = Number(s);
  return Number.isFinite(n) ? n : 0;
};
const asBool = (v) => {
  if (typeof v === "boolean") return v;
  const s = String(v).trim().toLowerCase();
  return s === "1" || s === "true" || s === "t" || s === "yes" || s === "y";
};

/** ---------- Tunable thresholds / weights ---------- */
const FLOAT_BUCKETS = [
  [5_000_000, 35],
  [10_000_000, 30],
  [20_000_000, 25],
  [50_000_000, 15],
  [Infinity, 5],
];

const REGSHO_BONUS = 4;

/** Guardrails (minimum baseline rules) */
function applyMinimumsAdditive(score, floatShares, ctb, siPct) {
  if (floatShares < 10_000_000 && ctb > 300) score = Math.max(score, 91);
  if (floatShares < 5_000_000 && ctb > 100) score = Math.max(score, 88);
  if (floatShares < 30_000_000 && ctb > 50 && siPct >= 20) score = Math.max(score, 85);
  return score;
}
function applyMinimumsMultiplicative(score, floatShares, ctb, siPct) {
  if (floatShares < 10_000_000 && ctb > 300) score = Math.max(score, 91);
  if (floatShares < 5_000_000 && ctb > 100) score = Math.max(score, 88);
  if (floatShares < 30_000_000 && ctb > 50 && siPct >= 20) score = Math.max(score, 85);
  return score;
}

/** ---------- Multipliers for multiplicative model ---------- */
function floatMultiplier(floatShares) {
  if (floatShares < 5_000_000) return 3.0;
  if (floatShares < 10_000_000) return 2.5;
  if (floatShares < 20_000_000) return 2.0;
  if (floatShares < 50_000_000) return 1.5;
  return 1.0;
}
function ctbMultiplier(ctb) {
  if (ctb >= 500) return 3.0;
  if (ctb >= 300) return 2.7;
  if (ctb >= 100) return 2.0;
  if (ctb >= 50)  return 1.6;
  if (ctb >= 25)  return 1.3;
  return 1.0;
}
function siMultiplier(siPct) {
  if (siPct >= 50) return 2.5;
  if (siPct >= 20) return 2.0;
  if (siPct >= 10) return 1.5;
  if (siPct >= 5)  return 1.2;
  return 1.0;
}
function ftdMultiplier(ftdVal) {
  if (ftdVal >= 10_000_000) return 1.3;
  if (ftdVal >= 5_000_000)  return 1.2;
  if (ftdVal >= 1_000_000)  return 1.1;
  return 1.0;
}
const regshoMultiplier = (regsho) => (regsho ? 1.05 : 1.0);

/** ---------- Additive model ---------- */
export function scoreAdditive({ float_shares, si_pct, ctb, ftd_val, regsho }) {
  let score = 0;

  // Float
  const floatPts = FLOAT_BUCKETS.find(([thr]) => float_shares < thr)?.[1] ?? 5;
  score += floatPts;

  // CTB
  if (ctb >= 500) score += 24;
  else if (ctb >= 300) score += 23;
  else if (ctb >= 100) score += 19;
  else if (ctb >= 50)  score += 14;
  else if (ctb >= 25)  score += 10;
  else if (ctb >= 10)  score += 5;

  // SI%
  if (si_pct >= 50) score += 25;
  else if (si_pct >= 20) score += 19;
  else if (si_pct >= 10) score += 14;
  else if (si_pct >= 5)  score += 10;

  // Bonuses
  if (float_shares < 10_000_000 && si_pct >= 20) score += 8;
  if (float_shares < 5_000_000 && ctb > 100)     score += 12;
  if (float_shares < 10_000_000 && ctb > 300)    score += 8;

  // FTDs ($)
  if (ftd_val >= 10_000_000)      score += 9;
  else if (ftd_val >= 5_000_000)   score += 7;
  else if (ftd_val >= 1_000_000)   score += 5;

  // RegSHO bonus only
  if (regsho) score += REGSHO_BONUS;

  // Guardrails & cap
  score = applyMinimumsAdditive(score, float_shares, ctb, si_pct);
  return clamp99(score);
}

/** ---------- Multiplicative model ---------- */
export function scoreMultiplicative({ float_shares, si_pct, ctb, ftd_val, regsho }) {
  const raw =
    floatMultiplier(float_shares) *
    ctbMultiplier(ctb) *
    siMultiplier(si_pct) *
    ftdMultiplier(ftd_val) *
    regshoMultiplier(regsho);

  // Normalize to ~0–100; 45 is empirical “max” anchor
  let score = (raw / 45) * 100;

  // Guardrails & cap
  score = applyMinimumsMultiplicative(score, float_shares, ctb, si_pct);
  return clamp99(score);
}

/** ---------- Consensus ---------- */
export function scoreConsensus({ float_shares, os_shares = 0, si_pct, ctb, ftd_val, regsho }) {
  const add = scoreAdditive({ float_shares, si_pct, ctb, ftd_val, regsho });
  const mul = scoreMultiplicative({ float_shares, si_pct, ctb, ftd_val, regsho });
  return Math.round(((add + mul) / 2) * 10) / 10;
}

/** ---------- Breakdown (for debugging) ---------- */
export function scoreBreakdown({ float_shares, os_shares = 0, si_pct, ctb, ftd_val, regsho }) {
  // Additive component points
  const floatPts = FLOAT_BUCKETS.find(([thr]) => float_shares < thr)?.[1] ?? 5;
  const ctbPts =
    ctb >= 500 ? 24 :
    ctb >= 300 ? 23 :
    ctb >= 100 ? 19 :
    ctb >= 50  ? 14 :
    ctb >= 25  ? 10 :
    ctb >= 10  ? 5  : 0;
  const siPts =
    si_pct >= 50 ? 25 :
    si_pct >= 20 ? 19 :
    si_pct >= 10 ? 14 :
    si_pct >= 5  ? 10 : 0;

  let bonuses = 0;
  if (float_shares < 10_000_000 && si_pct >= 20) bonuses += 8;
  if (float_shares < 5_000_000 && ctb > 100)     bonuses += 12;
  if (float_shares < 10_000_000 && ctb > 300)    bonuses += 8;

  const ftdPts =
    ftd_val >= 10_000_000 ? 9 :
    ftd_val >= 5_000_000  ? 7 :
    ftd_val >= 1_000_000  ? 5 : 0;

  const regPts = regsho ? REGSHO_BONUS : 0;

  let additiveTotal = floatPts + ctbPts + siPts + bonuses + ftdPts + regPts;
  additiveTotal = applyMinimumsAdditive(additiveTotal, float_shares, ctb, si_pct);
  additiveTotal = clamp99(additiveTotal);

  const multiplicativeTotal = scoreMultiplicative({ float_shares, si_pct, ctb, ftd_val, regsho });
  const consensus = Math.round(((additiveTotal + multiplicativeTotal) / 2) * 10) / 10;

  return {
    Float: floatPts,
    CTB: ctbPts,
    "SI%": siPts,
    Bonuses: bonuses,
    FTD: ftdPts,
    RegSHO: regPts,
    "Additive Total": Number(additiveTotal.toFixed(1)),
    "Multiplicative Total": Number(multiplicativeTotal.toFixed(1)),
    Consensus: consensus,
  };
}

/** ---------- Batch helpers ---------- */
export function scoreMany(rows) {
  return rows.map((r) =>
    Object.assign(
      { ticker: r.ticker ?? "" },
      scoreBreakdown({
        float_shares: asFloat(r.float_shares),
        os_shares: asFloat(r.os_shares),
        si_pct: asFloat(r.si_pct),
        ctb: asFloat(r.ctb),
        ftd_val: asFloat(r.ftd_val),
        regsho: asBool(r.regsho),
      })
    )
  );
}

/** ---------- CSV parsing (robust) ---------- */
const HEADER_ALIASES = {
  ticker: new Set(["ticker", "symbol"]),
  float_shares: new Set(["float_shares", "free float", "float", "free_float"]),
  os_shares: new Set(["os_shares", "shares outstanding", "outstanding", "os", "shares_outstanding"]),
  si_pct: new Set(["si_pct", "short float %", "short %", "short_percent", "short_float_pct"]),
  ctb: new Set(["ctb", "cost to borrow", "borrow cost", "ctb%", "cost_to_borrow"]),
  ftd_val: new Set(["ftd_val", "ftd $", "ftd value", "ftd", "ftd_usd"]),
  regsho: new Set(["regsho", "reg sho", "threshold", "on_regsho"]),
};

function normalizeHeaders(headers) {
  const lower = headers.map((h) => h.trim()).map((h) => [h.toLowerCase(), h]);
  const map = {};
  for (const [canon, aliases] of Object.entries(HEADER_ALIASES)) {
    for (const [low, raw] of lower) {
      if (aliases.has(low)) { map[canon] = raw; break; }
    }
  }
  return map;
}

function sniffDelimiter(firstLine) {
  if ((firstLine.match(/,/g) || []).length >= 1) return ",";
  if ((firstLine.match(/;/g) || []).length >= 1) return ";";
  if ((firstLine.match(/\t/g) || []).length >= 1) return "\t";
  return ","; // default
}

/**
 * Parse CSV text and return breakdown rows.
 * @param {string} csvText
 * @param {{debug?: boolean}=} opts
 * @returns {Array<object>}
 */
export function scoreFromCSV(csvText, opts = {}) {
  // Strip BOM if present
  if (csvText.charCodeAt(0) === 0xFEFF) csvText = csvText.slice(1);
  const lines = csvText.split(/\r\n|\r|\n/);
  if (!lines.length) return [];
  const delim = sniffDelimiter(lines[0]);

  // Simple CSV splitter (no embedded quotes support). For robust CSVs use PapaParse.
  const split = (line) => line.split(delim).map((s) => s.trim());

  const header = split(lines[0]);
  const map = normalizeHeaders(header);
  const idx = (canon) => header.indexOf(map[canon] ?? "__missing__");

  const out = [];
  for (let i = 1; i < lines.length; i++) {
    const row = lines[i];
    if (!row.trim()) continue;
    const cols = split(row);
    try {
      const payload = {
        ticker: cols[idx("ticker")] ?? "",
        float_shares: asFloat(cols[idx("float_shares")]),
        os_shares: asFloat(cols[idx("os_shares")]),
        si_pct: asFloat(cols[idx("si_pct")]),
        ctb: asFloat(cols[idx("ctb")]),
        ftd_val: asFloat(cols[idx("ftd_val")]),
        regsho: asBool(cols[idx("regsho")]),
      };
      out.push(Object.assign({ ticker: payload.ticker }, scoreBreakdown(payload)));
    } catch (e) {
      if (opts.debug) console.warn(`[row ${i}] parse error:`, e, row);
    }
  }
  return out;
}

/** ---------- Example usage (in extension background or content script) ----------
import { scoreConsensus, scoreBreakdown, scoreMany, scoreFromCSV } from './squeezeScore.js';

// Single
const zone = scoreConsensus({ float_shares: 7_150_000, si_pct: 0.83, ctb: 755, ftd_val: 10_000_000, regsho: true });
console.log('ZONE consensus:', zone);

// Breakdown
console.log(scoreBreakdown({ float_shares: 3_540_000, si_pct: 37.67, ctb: 400, ftd_val: 1_620_000, regsho: true }));

// Batch
const rows = [
  { ticker: 'ZONE', float_shares: 7150000, si_pct: 0.83, ctb: 755, ftd_val: 10000000, regsho: true },
  { ticker: 'CWD',  float_shares: 3540000, si_pct: 37.67, ctb: 400, ftd_val: 1620000,  regsho: true },
];
console.table(scoreMany(rows));

// CSV
// const result = scoreFromCSV(csvText, { debug: true });
// console.table(result);
--------------------------------------------------------------------------- */
